dashboard/app/controllers/registrations_controller.rb

```diff
@@ -51,6 +55,29 @@ def create
     end
   end
 
+  def upgrade
+    return head(:bad_request) if params[:user].nil?
+    @user = User.find(current_user.id)
+    user_params = params[:user]
+
+    can_update =
+      if @user.teacher_managed_account?
+        if @user.secret_word_account?
+          secret_words_match = user_params[:secret_words] == @user.secret_words
+          error_string = user_params[:secret_words].blank? ? :blank_plural : :invalid_plural
+          @user.errors.add(:secret_words, error_string) unless secret_words_match
+          secret_words_match
+        else
+          true
+        end
+      else
+        false
+      end
+
+    successfully_updated = can_update && @user.update!(update_params(params))
```

Comment:
`update!` (the bang-method) will raise an error if the update fails. Here it will cause a server error. The non-bang version, `update` will return a bool whether the update succeeded or not. That seems like what you want here.